maxfw/utils/image_functions.py

```diff
@@ -0,0 +1,443 @@
+            # if 2D image, add channel dimension (HWC)
+            pic = np.expand_dims(pic, 2)
+
+    elif isinstance(pic, (bytes, bytearray)):
+        try:
+            # verify that the object can be loaded into memory
+            pic = np.array(Image.open(io.BytesIO(pic)))
+        except Exception:
+            raise TypeError('The input bytes object is not suitable for the Pillow library. Check the input again.')
+
+    npimg = pic
+    if not isinstance(npimg, np.ndarray):
+        raise TypeError('Input pic must be a bytes object or NumPy ndarray, ' +
+                        'not {}'.format(type(npimg)))
+
+    if npimg.shape[2] == 1:
+        expected_mode = None
+        npimg = npimg[:, :, 0]
+        if npimg.dtype == np.uint8:
+            expected_mode = 'L'
+        elif npimg.dtype == np.int16:
+            expected_mode = 'I;16'
+        elif npimg.dtype == np.int32:
+            expected_mode = 'I'
+        elif npimg.dtype == np.float32:
+            expected_mode = 'F'
+        if mode is not None and mode != expected_mode:
+            raise ValueError("Incorrect mode ({}) supplied for input type {}. Should be {}"
+                             .format(mode, np.dtype, expected_mode))
+        mode = expected_mode
+
+    elif npimg.shape[2] == 2:
+        permitted_2_channel_modes = ['LA']
+        if mode is not None and mode not in permitted_2_channel_modes:
+            raise ValueError("Only modes {} are supported for 2D inputs".format(permitted_2_channel_modes))
+
+        if mode is None and npimg.dtype == np.uint8:
+            mode = 'LA'
+
+    elif npimg.shape[2] == 4:
+        permitted_4_channel_modes = ['RGBA', 'CMYK', 'RGBX']
+        if mode is not None and mode not in permitted_4_channel_modes:
+            raise ValueError("Only modes {} are supported for 4D inputs".format(permitted_4_channel_modes))
+
+        if mode is None and npimg.dtype == np.uint8:
+            mode = 'RGBA'
+    else:
+        permitted_3_channel_modes = ['RGB', 'YCbCr', 'HSV']
+        if mode is not None and mode not in permitted_3_channel_modes:
+            raise ValueError("Only modes {} are supported for 3D inputs".format(permitted_3_channel_modes))
+        if mode is None and npimg.dtype == np.uint8:
+            mode = 'RGB'
+
+    if mode is None:
+        raise TypeError('Input type {} is not supported'.format(npimg.dtype))
+
+    # Verify that the target mode exists
+    assert target_mode in [1, 'L', 'P', 'RGB', 'RGBA', 'CMYK', 'YCbCr', 'LAB', 'HSV', 'I', 'F', 'RGBX', 'RGBBa']
+
+    return Image.fromarray(npimg, mode=mode).convert(target_mode)
+
+
+def normalize(img):
+    if type(img) is not np.ndarray:
+        img = np.array(img)
+    img = img / (np.max(img) - np.min(img))
+    return Image.fromarray(img.astype('uint8'))
+
+
+def standardize(img):
+    if type(img) is not np.ndarray:
+        img = np.array(img)
+    mean = np.mean(img)
+    std = np.std(img)
+    return (img - mean) / std
```

Comment:
This should be returning a PIL image, right?

Reply:
Actually, I have taken a deeper look at this and I don't think it's possible for it to return a PIL image object. The conversion to a PIL object for the `Normalize` function is an error as well I believe. The explanation for this is that Pillow objects can not be generated from matrices with floats and/or negative numbers, and we most certainly need floats/negatives to represent a normalized or standardized image. 

The way we got this to not throw an error in the `normalize` function is because we changed the dtype to `uint8`, which mistakenly converted negatives and floats to integers.

Because we can not convert the output of this transformation to a PIL object, we can not connect the output to e.g.  a rotate or resize functionality. This means that the `standardize` or `normalize` transforms should be the last operation in the pipeline. In order to enforce this I have provided an assert statement in the next update.

Thoughts on this @MLnick @kmh4321?

Reply:
Yes, since this is a purely matrix related operation, this would be done outside the scope of PIL operations (like rotate, crop  etc). Even Torch vision transforms implements this as pure pytorch tensors while the other operations still handle/work on PIL Image objects. That being said, we will document this clearly and state the assumption that these operations go last. 
@splovyt @MLnick